File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> traversal;
        if (root == nullptr) return traversal;
        queue<TreeNode*> q;
        q.push(root);
        bool l_to_r = true;
        while(!q.empty()){
            int n = q.size();
            vector<int> level(n, 0);
            for(int i=0; i<n; i++){
                TreeNode *node = q.front();
                q.pop();

                int idx = (l_to_r) ? i : n-i-1;

                level[idx] = node->val;

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            traversal.push_back(level);
            l_to_r = !l_to_r;
        }

        return traversal;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
    var traversal [][]int
    if root == nil {
        return traversal
    }

    var q []*TreeNode
    q = append(q, root)
    flag := true

    for len(q) != 0 {
        n := len(q)
        var level []int

        for i := 0; i < n; i++ {
            node := q[0]
            q = q[1:]

            level = append(level, node.Val)

            if node.Left != nil { q = append(q, node.Left) }
            if node.Right != nil { q = append(q, node.Right) }
        }

        if flag == false {
            reverse(level)
        }

        flag = !flag
        traversal = append(traversal, level)
    }

    return traversal
}
func reverse(nums []int) {
    i, j := 0, len(nums)-1
    for i < j {
        nums[i], nums[j] = nums[j], nums[i]
        i++
        j--
    }
}